[media-controls] Duplicate buttons in top leading corner of fullscreen video

pxlcoder opened the pull request:
#### 5d182d0202ae63407fec2fcb18bde76f269dc021
<pre>
[media-controls] Duplicate buttons in top leading corner of fullscreen video
<a href="https://bugs.webkit.org/show_bug.cgi?id=246414">https://bugs.webkit.org/show_bug.cgi?id=246414</a>
rdar://100870340

Reviewed by Devin Rousso.

254333@main attempted to hide the hide the cancel and PiP buttons in fullscreen
if the current playback element already has those buttons. However, at the time
the `EnterFullScreen` message in sent to the UIProcess, the `PlaybackSessionManager`
may not have a current playback element.

To fix, explicitly check if the element entering fullscreen is a `&lt;video&gt;` and
that it has controls. Additionally, rename `isVideoElement` to
`isVideoElementWithControls` for accuracy.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::isVideoElementWithControls const):
(WebKit::WebFullScreenManagerProxy::enterFullScreen):
(WebKit::WebFullScreenManagerProxy::isVideoElement const): Deleted.
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.cpp:
(WebKit::InjectedBundlePageFullScreenClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.h:

Canonical link: <a href="https://commits.webkit.org/255503@main">https://commits.webkit.org/255503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a713907e41abb22a815eb3959693141b8eb2230

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92590 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1809 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23175 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102328 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1807 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30168 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84993 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98491 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98253 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1216 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79086 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28152 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83134 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82835 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71243 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36578 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16762 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34364 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17941 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3813 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38234 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40554 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40143 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37100 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -171,11 +171,13 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element)
-    bool isVideoElement = false;
+    bool isVideoElementWithControls = false;
 #if PLATFORM(IOS_FAMILY) || (PLATFORM(MAC) && ENABLE(VIDEO_PRESENTATION_MODE))
+    if (auto* videoElement = dynamicDowncast<HTMLVideoElement>(element))
+        isVideoElementWithControls = videoElement->controls();
```
should we have this be an `else` in case we do have a `currentPlaybackControlsElement` (or if it doesn't match the `element`)?

dcrousso replied:
e.g. is it possible for the `currentPlaybackControlsElement` to be a different `<video>` (or even an `<audio>`)?

pxlcoder replied:
> should we have this be an else in case we do have a currentPlaybackControlsElement (or if it doesn't match the element)?

I don't think so, the intent is to hide the controls on the UIProcess-side, if the element being full screened (`element`) already has controls. I believe the logic I have captures that case.

The duplicated controls are specific to `video` elements, so we do not need to handle `audio`.

dcrousso commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -171,11 +171,13 @@ void WebFullScreenManager::enterFullScreenForElement(WebCore::Element* element)
+    if (auto* videoElement = dynamicDowncast<HTMLVideoElement>(element))
+        isVideoElementWithControls = videoElement->controls();
+
     if (auto* currentPlaybackControlsElement = m_page->playbackSessionManager().currentPlaybackControlsElement()) {
```
Style: no `{` `}` for single line control flow

pxlcoder committed:
[media-controls] Duplicate buttons in top leading corner of fullscreen video
https://bugs.webkit.org/show_bug.cgi?id=246414
rdar://100870340

Reviewed by Devin Rousso.

254333@main attempted to hide the hide the cancel and PiP buttons in fullscreen
if the current playback element already has those buttons. However, at the time
the `EnterFullScreen` message in sent to the UIProcess, the `PlaybackSessionManager`
may not have a current playback element.

To fix, explicitly check if the element entering fullscreen is a `<video>` and
that it has controls. Additionally, rename `isVideoElement` to
`isVideoElementWithControls` for accuracy.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::isVideoElementWithControls const):
(WebKit::WebFullScreenManagerProxy::enterFullScreen):
(WebKit::WebFullScreenManagerProxy::isVideoElement const): Deleted.
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.cpp:
(WebKit::InjectedBundlePageFullScreenClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.h:

Canonical link: https://commits.webkit.org/255503@main
